tools/binpac/src/pac_analyzer.cc

```diff
@@ -46,45 +46,46 @@ AnalyzerDecl::~AnalyzerDecl() {
 
 void AnalyzerDecl::AddElements(AnalyzerElementList* elemlist) {
     ASSERT(! env_);
-    foreach (i, AnalyzerElementList, elemlist) {
-        AnalyzerElement* elem = *i;
-        switch ( elem->type() ) {
-            case AnalyzerElement::STATE: {
-                ASSERT(0);
-                AnalyzerState* state_elem = (AnalyzerState*)elem;
-                statevars_->insert(statevars_->end(), state_elem->statevars()->begin(), state_elem->statevars()->end());
-            } break;
-            case AnalyzerElement::ACTION: {
-                ASSERT(0);
-                AnalyzerAction* action_elem = (AnalyzerAction*)elem;
-                actions_->push_back(action_elem);
-            } break;
-            case AnalyzerElement::HELPER: {
-                AnalyzerHelper* helper_elem = (AnalyzerHelper*)elem;
-
-                switch ( helper_elem->helper_type() ) {
-                    case AnalyzerHelper::INIT_CODE: constructor_helpers_->push_back(helper_elem); break;
-                    case AnalyzerHelper::CLEANUP_CODE: destructor_helpers_->push_back(helper_elem); break;
-                    case AnalyzerHelper::EOF_CODE: eof_helpers_->push_back(helper_elem); break;
-                    default: helpers_->push_back(helper_elem);
-                }
-            } break;
-            case AnalyzerElement::FUNCTION: {
-                AnalyzerFunction* func_elem = (AnalyzerFunction*)elem;
-                Function* func = func_elem->function();
-                func->set_analyzer_decl(this);
-                functions_->push_back(func);
-            } break;
-            case AnalyzerElement::FLOW: {
-                AnalyzerFlow* flow_elem = (AnalyzerFlow*)elem;
-                ProcessFlowElement(flow_elem);
-            } break;
-            case AnalyzerElement::DATAUNIT: {
-                AnalyzerDataUnit* dataunit_elem = (AnalyzerDataUnit*)elem;
-                ProcessDataUnitElement(dataunit_elem);
-            } break;
+    if ( elemlist )
```

Comment:
This would've avoided the extra indentation and made the diff a bit smaller, but it's a nit.

```
if ( ! elemlist )
    return;
```

Reply:
Hmm, hmm, looking at the "Remove foreach macro" commit, I'd probably opt for ditching that commit and keep the macro. Seems that was an established pattern in BinPAC and changing appears mostly churn without a real upside. Totally IMO and feel free to ignore if you strictly prefer the no-macro version.

Reply:
How about making a templated version that's roughly identical to https://en.cppreference.com/w/cpp/algorithm/for_each.html, but checks for pointer validity first.

Reply:
I think something like this should work:

```
template<typename T>
concept Container = requires(T t) {
    std::begin(t);
    std::end(t);
};

template<Container C, class Pred>
constexpr void ptr_for_each(C* c, Pred f) {
    if ( ! c )
        return;

    std::ranges::for_each(*c, f);
}
```

it wouldn't work in every case, such as ones where the loop returns from the caller function (such as [this one](https://github.com/zeek/zeek/pull/4782/commits/b068d870f7a1ad847be7c5f071fc23530546f6e3#diff-2576e9ef4f844a79307e475e39e9531f7b2cea72c260188fc469a05f270b8033R46-R48)), but it should work in most of the cases.

Reply:
> How about making a templated version that's roughly identical to

I think you need someone else to chime in on that :-)

It doesn't seem that BinPAC's `foreach` macro can easily escape into Zeek's code due to the separation between `src/pac_` files and `lib/binpac.h`. And then seems just a detail in BinPAC's implementation that's been like that for years and IMO would certainly justify a `// NOLINT` annotation. If it ends up something else, great, but don't have enough opinions or inspiration on that something else (the open-coding just stood out because of the churn and the `foreach` appeared pragmatic to me, but that might just be me).

Reply:
Makes sense. I dropped that commit out.